### PR TITLE
support php 7.4 field type hints in instance serialization

### PR DESCRIPTION
**Pull request by unserialize**
Before this MR, we parsed `@var` for fields marked as `@kphp-serialized-field`.
Now, we support PHP 7.4 field type hints:
```
  /** @kphp-serialized-field 1 */
  public ?int $i_f = null;
```
If a field has a type hint, construct it as a string and parse with a regular phpdoc parser, like it was written in `@var`.

Relates to https://github.com/VKCOM/kphp/pull/139

### CODE REVIEW

File: src/InstanceSerialization/InstanceMetadata.php [hunk 1]
```diff
@@ -92,13 +92,22 @@ public function __construct(string $instance) {
...
       $this->as_float32[] = strpos($property->getDocComment(), '@kphp-serialized-float32') !== false;
 
+      // get type either from @var or from php 7.4 field type hint
+      $type = '';
       preg_match('/@var\s+([^\n]+)/', $property->getDocComment(), $matches);
-      assert(count($matches) > 1);
-      $type     = (string)$matches[1];
+      if (count($matches) > 1) {
+        $type = (string)$matches[1];
+      } else if (version_compare(phpversion(), '7.4', '>=') && $property->hasType()) {
```
Comment (rybakit) [inline src/InstanceSerialization/InstanceMetadata.php:100]: Can be simplified to `PHP_VERSION_ID >= 70400`.

File: src/InstanceSerialization/InstanceMetadata.php [hunk 2]
```diff
@@ -92,13 +92,22 @@ public function __construct(string $instance) {
 
       $this->as_float32[] = strpos($property->getDocComment(), '@kphp-serialized-float32') !== false;
 
+      // get type either from @var or from php 7.4 field type hint
+      $type = '';
       preg_match('/@var\s+([^\n]+)/', $property->getDocComment(), $matches);
-      assert(count($matches) > 1);
-      $type     = (string)$matches[1];
+      if (count($matches) > 1) {
+        $type = (string)$matches[1];
```
Comment (rybakit) [inline src/InstanceSerialization/InstanceMetadata.php:99]: Casting to string is not necessary, all `$mathes` values are strings.

File: src/InstanceSerialization/InstanceMetadata.php [hunk 3]
```diff
@@ -92,13 +92,22 @@ public function __construct(string $instance) {
...
+      } else if (version_compare(phpversion(), '7.4', '>=') && $property->hasType()) {
+        $type = ($property->getType()->allowsNull() ? '?' : '') . (string)$property->getType();
```
Comment (rybakit) [inline src/InstanceSerialization/InstanceMetadata.php:101]: Casting to string is not necessary, the `.` concatenation casts the argument.

File: src/InstanceSerialization/InstanceMetadata.php [hunk 4]
```diff
@@ -92,13 +92,22 @@ public function __construct(string $instance) {
...
+      }
+      if ($type === '' || $type === null) {
```
Comment (rybakit) [inline src/InstanceSerialization/InstanceMetadata.php:103]: `$type` can't be `null`?

File: src/InstanceSerialization/InstanceMetadata.php (hunk 1 above)
Comment (unserialize) [reply src/InstanceSerialization/InstanceMetadata.php:100]: +

File: src/InstanceSerialization/InstanceMetadata.php (hunk 2 above)
Comment (unserialize) [reply src/InstanceSerialization/InstanceMetadata.php:99]: Yes, but... Whyever, PhpStorm doesn't think so in its internals 😀
image
So, I'll leave this cast for now

**Review by AlexK0 — APPROVED** [review]

**Pull request merged**